### DISCUSSION TITLE
lazy NaN

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#7057] [Q&A]
Hello,

For my new R package **lazyNumbers** I want to be able to deal with `NaN` values.

I'm able to create a lazy `NaN` by doing `lazyNumber(0) / lazyNumber(0)`. Then I indeed get `NaN` when I call `as_double` on this lazy number.

But how to test whether a lazy number is `NaN`? I could test whether its double approximation is `NaN` but this is not a good solution, I don't want to call `as_double` everytime.

I already deal with *missing* values by using `std::optional`. So I cannot use `std::optional` for the `NaN` values.

**Comment by stla** [comment]
Hmm I've just have an idea: a lazy number is `NaN` if and only if its enclosing interval is `]-Inf, Inf[`. Is it a reasonable way?

**Reply by stla** [reply]
> a lazy number is `NaN` if and only if its enclosing interval is `]-Inf, Inf[`. Is it a reasonable way?

Wrong. The enclosing interval of an infinite lazy number is `]-Inf, Inf[` as well. 

This is unexpected to me: why the interval of the positive infinite lazy number is not `inf = sup = Inf`?

**Reply by stla** [reply]
*Updated idea*: a lazy number `x` is `NaN` if and only if the enclosing interval of `x` is `]-Inf, Inf[` and the enclosing interval of `1/x` is `]-Inf, Inf[`. Is it correct?

**Reply by MaelRL** [reply]
> > a lazy number is `NaN` if and only if its enclosing interval is `]-Inf, Inf[`. Is it a reasonable way?
> 
> Wrong. The enclosing interval of an infinite lazy number is `]-Inf, Inf[` as well.
> 
> This is unexpected to me: why the interval of the positive infinite lazy number is not `inf = sup = Inf`?

In `CGAL::Interval_nt`, `_inf` is actually `- inf`, not `inf`. You should access the inf bound with `inf()`, and not look at `_inf`.

**Reply by stla** [reply]
> _Updated idea_: a lazy number `x` is `NaN` if and only if the enclosing interval of `x` is `]-Inf, Inf[` and the enclosing interval of `1/x` is `]-Inf, Inf[`. Is it correct?

Wrong again :-( Sounds strange: if `x` is `lazyNumber(Inf)`, then the enclosing interval of `1/x` is `]-Inf, Inf[` (while the double approximation of `x` is as expected `0`).

**Comment by MaelRL** [comment]
Each number type `NT` in CGAL has a specialization `Is_valid<NT>`, e.g.:
```
template< class ET >
class Is_valid< Lazy_exact_nt<ET> >
  : public CGAL::cpp98::unary_function< Lazy_exact_nt<ET>, bool > {
  public :
    bool operator()( const Lazy_exact_nt<ET>& x ) const {
      return is_valid(x.approx());
    }
};
```
which will check if the number is valid (which includes not being a NaN). Could you use this?

**Reply by stla** [reply]
Thanks. I've just tried `CGAL::is_valid(x)` and this returns `true` for `x = lazyNumber(0) / lazyNumber(0)`.

**Comment by stla** [comment]
Still attempted with no luck. My conclusion is that there's no way without using the double approximation.
